File: iOS/SayHi/SayHi/Message.swift
//
//  Match.swift
//  SayHi
//
//  Created by Klemenz, Oliver on 12.12.16.
//  Copyright © 2016 Klemenz, Oliver. All rights reserved.
//

import Foundation

struct Message {

    static let MatchPartCount = 13
    static let PartSeparatorChracter = "\0"
    static let TagSeparatorChracter = "\t"
    
    var langCode: String
    var firstName: String
    var gender: Gender
    var birthYear: Int
    var status: String
    var space: String
    var installationUUID: String
    var matchMode: MatchMode
    var matchHandshake: Bool
    var posTags : [String] = []
    var negTags : [String] = []
    
    var description : String {
        let parts : [String] = [
            UserData.instance.langCode,
            firstName.condensed,
            gender.rawValue,
            String(birthYear),
            status.condensed,
            space.condensed,
            installationUUID,
            String(matchMode.rawValue),
            String(matchHandshake ? 1 : 0),
            String(posTags.count),
            posTags.joined(separator: Message.TagSeparatorChracter),
            String(negTags.count),
            negTags.joined(separator: Message.TagSeparatorChracter)
        ]
        return parts.joined(separator: Message.PartSeparatorChracter)
    }
    
    static func generate(profile: Profile) -> Message {
        let userData = UserData.instance
        return Message(langCode: userData.langCode,
                       firstName: userData.firstName,
                       gender: userData.gender,
                       birthYear: userData.birthYear,
                       status: userData.status,
                       space: SecureStore.spaceRefName,
                       installationUUID: userData.installationUUID,
                       matchMode: profile.effectiveMatchMode,
                       matchHandshake: userData.matchHandshake,
                       posTags: profile.posTagEffectiveKeys,
                       negTags: profile.negTagEffectiveKeys)
    }
    
    static func parse(_ text: String) -> Message? {
        let parts = text.components(separatedBy: PartSeparatorChracter)
        if parts.count == MatchPartCount {
            let langCode = parts[0]
            let firstName = parts[1]
            let gender = Gender(rawValue: parts[2]) ?? .none
            let birthYear = Int(parts[3]) ?? BaseYear
            let status = parts[4]
            let space = parts[5]
            let installationUUID = parts[6]
            let matchMode = MatchMode(rawValue: Int(parts[7]) ?? MatchMode.open.rawValue) ?? .open
            let matchHandshake = Int(parts[8]) == 1
            let _ = Int(parts[9]) ?? 0
            let posTags = !parts[10].isEmpty ? parts[10].components(separatedBy: TagSeparatorChracter) : []
            let _ = Int(parts[11]) ?? 0
            let negTags = !parts[12].isEmpty ? parts[12].components(separatedBy: TagSeparatorChracter) : []
            
            return Message(langCode: langCode,
                           firstName: firstName,
                           gender: gender,
                           birthYear: birthYear,
                           status: status,
                           space: space,
                           installationUUID: installationUUID,
                           matchMode: matchMode,
                           matchHandshake: matchHandshake,
                           posTags: posTags,
                           negTags: negTags)
        }
        return nil
    }
}

File: Android/SayHi/app/src/main/java/de/oklemenz/sayhi/model/Message.java
package de.oklemenz.sayhi.model;

import android.text.TextUtils;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import de.oklemenz.sayhi.service.SecureStore;
import de.oklemenz.sayhi.service.Utilities;

/**
 * Created by Oliver Klemenz on 22.02.17.
 */

public class Message implements Serializable {

    private static int MatchPartCount = 13;
    private static String PartSeparatorCharacter = "\0";
    private static String TagSeparatorCharacter = "\t";

    public String langCode;
    public String firstName;
    public Enum.Gender gender;
    public int birthYear;
    public String status;
    public String space;
    public String installationUUID;
    public Enum.MatchMode matchMode;
    public boolean matchHandshake;
    public List<String> posTags = new ArrayList<>();
    public List<String> negTags = new ArrayList<>();

    public Message(String langCode, String firstName, Enum.Gender gender, int birthYear, String status, String space, String installationUUID, Enum.MatchMode matchMode, boolean matchHandshake, List<String> posTags, List<String> negTags) {
        this.langCode = langCode;
        this.firstName = firstName;
        this.gender = gender;
        this.birthYear = birthYear;
        this.status = status;
        this.space = space;
        this.installationUUID = installationUUID;
        this.matchMode = matchMode;
        this.matchHandshake = matchHandshake;
        this.posTags = posTags;
        this.negTags = negTags;
    }

    @Override
    public String toString() {
        String[] parts = {
                langCode,
                Utilities.condense(firstName),
                gender.code,
                Integer.toString(birthYear),
                Utilities.condense(status),
                Utilities.condense(space),
                installationUUID,
                Integer.toString(matchMode.code),
                Integer.toString(matchHandshake ? 1 : 0),
                Integer.toString(posTags.size()),
                TextUtils.join(TagSeparatorCharacter, posTags),
                Integer.toString(negTags.size()),
                TextUtils.join(TagSeparatorCharacter, negTags)
        };
        return TextUtils.join(PartSeparatorCharacter, parts);
    }

    public static Message generate(Profile profile) {
        UserData userData = UserData.getInstance();
        return new Message(
                userData.getLangCode(),
                userData.getFirstName(),
                userData.getGender(),
                userData.getBirthYear(),
                userData.getStatus(),
                SecureStore.getSpaceRefName(),
                userData.getInstallationUUID(),
                profile.getEffectiveMatchMode(),
                userData.getMatchHandshake(),
                profile.getPosTagEffectiveKeys(),
                profile.getNegTagEffectiveKeys()
        );
    }

    public static Message parse(String text) {
        String[] parts = text.split(PartSeparatorCharacter, -1);
        if (parts.length == MatchPartCount) {
            String langCode = parts[0];
            String firstName = parts[1];
            Enum.Gender gender = Enum.Gender.fromCode(parts[2]);
            int birthYear = Integer.parseInt(parts[3]) >= UserData.BaseYear ? Integer.parseInt(parts[3]) : 0;
            String status = parts[4];
            String space = parts[5];
            String installationUUID = parts[6];
            Enum.MatchMode matchMode = Enum.MatchMode.fromCode(Integer.parseInt(parts[7]));
            boolean matchHandshake = Integer.parseInt(parts[8]) == 1;
            int posTagCount = Integer.parseInt(parts[9]);
            List<String> posTags = Arrays.asList(!TextUtils.isEmpty(parts[10]) ? parts[10].split(TagSeparatorCharacter) : new String[]{});
            int negTagCount = Integer.parseInt(parts[11]);
            List<String> negTags = Arrays.asList(!TextUtils.isEmpty(parts[12]) ? parts[12].split(TagSeparatorCharacter) : new String[]{});

            return new Message(
                    langCode,
                    firstName,
                    gender,
                    birthYear,
                    status,
                    space,
                    installationUUID,
                    matchMode,
                    matchHandshake,
                    posTags,
                    negTags
            );
        }
        return null;
    }
}